### PR TITLE
perf: emit events before hook calls

### PR DESCRIPTION
**Pull request by smol-ninja**
Closes https://github.com/sablier-labs/v2-core/issues/939.

### CODE REVIEW

**Comment by andreivladbrg** [comment]
@smol-ninja let's not udpate yet the precompiles, will do it after we address all the findings

**Review by andreivladbrg — CHANGES_REQUESTED** [review]
Since the only place where the `updateMetadata` modifier is used is in the `_update` hook, why not remove it completely and emit the event individually in the function? This approach is also more gas efficient

**Comment by smol-ninja** [comment]
Agree. Changing it right away.

**Comment by smol-ninja** [comment]
All changes done. You can review it now @andreivladbrg.

File: src/abstracts/SablierV2Lockup.sol [hunk 1]
```diff
@@ -376,6 +363,9 @@ abstract contract SablierV2Lockup is
         // Effects and Interactions: make the withdrawal.
         _withdraw(streamId, to, amount);
 
+        // Emits an ERC-4906 event to trigger an update of the NFT metadata.
```
Comment (PaulRBerg) [inline src/abstracts/SablierV2Lockup.sol:366]: Let's say just `Emit` here.

**Review by PaulRBerg — APPROVED** [review]
just one comment, otherwise LGTM

**Comment by smol-ninja** [comment]
@andreivladbrg looks good to merge?

**Review by andreivladbrg — APPROVED** [review]
yes ser

image

**Pull request merged**